### PR TITLE
add post about kallax

### PR DESCRIPTION
**Pull request by erizocosmico**
@Serabe @dpordomingo tell me if you think I missed something important that should be mentioned in the post, even if it is mentioned already in the README.

### CODE REVIEW

File: content/post/kallax.md [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+func (q *UserQuery) FindByUsernameLike(pattern string) *UserQuery {
+        return q.Where(kallax.Like(Schema.User.Username, pattern))
+}
+```
+
+**WARNING:** define these in your own model file, never on the `kallax.go`. You should never edit that file, as it will be nuked every time you regenerate your models.
+
+## Dealing with relationships
+
+Consider the following models. We have a people, which can have many pets and a pet has an inverse relationships with its owner.
```
Comment (Serabe) [inline content/post/kallax.md]: s/a people/people/ or s/a people/a person/

Comment (Serabe) [inline content/post/kallax.md]: s/inverse relationships/inverse relationship/

File: content/post/kallax.md [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+```go
+store := NewPersonStore(db)
+
+q := NewPersonQuery().
+        FindByName("Steven").
+        WithPets(nil) // you can also pass a condition if you want pets filtered
+person, err := store.FindOne(q)
+```
+
+Kallax generated methods in your query to automatically preload your relationships. Just use `With{FieldName}`.
```
Comment (Serabe) [inline content/post/kallax.md]: Don't understand the first phrase.

File: content/post/kallax.md [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+Note that relationships are **not** preloaded by default, they must be explicitly preloaded using such methods.
+**WARNING:** preloading retrieves **all** the models of the relationship matching the giving condition, or just all of them if none was given. If the N side of your 1:N relationship is really big you may want to be doing the query from the other side.
```
Comment (Serabe) [inline content/post/kallax.md]: s/models/records/

Comment (Serabe) [inline content/post/kallax.md]: s/you may want to be doing the query from the other side/you may want to query from the other side/

Still not like it, but it is a little clearer.

File: content/post/kallax.md [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+
+All 1:1 relationships are retrieved **in the same query** used to retrieve the main model, using JOINs. So, retrieving pets with their owners would result in **no extra queries** (just a more expensive one, but still faster as a result).
+
+One to many relationships are more complicated. The basic solution would be to retrieve a single model and then doing another query to retrieve all its relationships. But that is N+1, we needed something better than that.
+
+We solved that by doing **batching**. So we retrieve the main model in batches of N rows, then find the relationships of all these rows, merge them, and keep batching. For example, if the batches have a size of 50 (the default batch size, you can set it in the query with the `BatchSize` method) and we are retrieving 200 people. Instead of doing 201 queries, we only have to make 8. 4 for the 4 batches of people and 4 for retrieving the pets of all these batches. You might think this might be expensive, but the result is more than an order of magnitude faster than the other solution.
```
Comment (Serabe) [inline content/post/kallax.md]: s/you can set it/you can change it/

File: content/post/kallax.md (hunk 3 above)
Comment (erizocosmico) [inline content/post/kallax.md]: Any suggestion for rewording?

File: content/post/kallax.md [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+```go
+store := NewPersonStore(db)
+
+q := NewPersonQuery().
+        FindByName("Steven").
+        WithPets(nil) // you can also pass a condition if you want pets filtered
+person, err := store.FindOne(q)
+```
+
+Kallax generated methods on your query type to preload your the relationships of the model with the form `With{RelationshipFieldName}`.
```
Comment (Serabe) [inline content/post/kallax.md]: Remove _your_ after _preload_ and before _the relationships_.

File: content/post/kallax.md [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+Note that relationships are **not** preloaded by default, they must be explicitly preloaded using such methods.
+**WARNING:** preloading retrieves **all** the records of the relationship matching the giving condition, or just all of them if none was given. If the N side of your 1:N relationship is really big you may want to query from the other side.
+
+### The N+1 problem
+
+One of our goals was for kallax to be **as fast as possible**. Because of this, the naive `N+1` solution for retrieving relationships did not work for us.
+
+All 1:1 relationships are retrieved **in the same query** used to retrieve the main model, using JOINs. So, retrieving pets with their owners would result in **no extra queries** (just a more expensive one, but still faster as a result).
```
Comment (Serabe) [inline content/post/kallax.md]: Remove _retrieved/retrieve_ duplication.

File: content/post/kallax.md [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+There's still a long road ahead of us. Lots of performance improvements we can make, features to introduce, etc, but the ones that are closer in the roadmap might be the following.
+
+* Automatically generate SQL schema from your models.
+* Many to many relationships.
+* Migrations.
+
+## Conclusion
+
+We've had quite the journey developing kallax. We had clear goals and we feel like they've been successfully achieved. Reinvent the wheel is not good most of the time, but this time we were covering a need that was there. This is not just another ORM, this is an ORM that cares care of some specific needs.
+We are pretty happy with the result and the first benchmark we've run, and we will definitely keep improving it in the future, since we are starting to use it in production, which guarantees bugs will get fixed and more features will get added.
```
Comment (Serabe) [inline content/post/kallax.md]: If this is supposed to be a different paragraph, add a white line before.

**Comment by erizocosmico** [comment]
Done @Serabe

**Comment by margaridag** [comment]
Added some minor fixes in #84.

**Pull request merged**